### PR TITLE
[MCCAS] Add a new CAS representation for object directly from MC

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
WIP. No tests yet.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@benlangmuir @akyrtzi I think this one is ready to be landed and I can iterate future improvement in tree.

@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: clang/include/clang/Basic/CodeGenOptions.def [hunk 1]
```diff
@@ -193,6 +193,7 @@ CODEGENOPT(StrictFloatCastOverflow, 1, 1)
 
 CODEGENOPT(UniformWGSize     , 1, 0) ///< -cl-uniform-work-group-size
 CODEGENOPT(NoZeroInitializedInBSS , 1, 0) ///< -fno-zero-initialized-in-bss.
+
```
Comment (benlangmuir) [inline clang/include/clang/Basic/CodeGenOptions.def]: Nit: would be good to avoid unnecessary changes to minimize merge conflicts.

File: clang/include/clang/Frontend/Utils.h [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #include "llvm/ADT/StringMap.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/ADT/StringSet.h"
+#include "llvm/CAS/CASDB.h"
```
Comment (benlangmuir) [inline clang/include/clang/Frontend/Utils.h]: Are the changes to this file necessary for something?

File: clang/tools/driver/cc1_main.cpp [hunk 3]
```diff
@@ -491,10 +512,33 @@ Optional<int> CompileJobCache::replayCachedResult(llvm::cas::ObjectRef ResultID,
...
+        Optional<llvm::cas::NodeProxy> CASObj;
+        if (Error E = CAS->getNode(BytesID).moveInto(CASObj))
+          llvm::report_fatal_error(std::move(E));
+        auto Schema =
+            std::make_unique<llvm::mccasformats::flatv1::MCSchema>(*CAS);
+        if (auto E = Schema->serializeObjectFile(*CASObj, OS))
+          report_fatal_error(std::move(E));
+      } else {
+        Optional<llvm::cas::BlobProxy> Bytes;
+        if (Error E = CAS->getBlob(BytesID).moveInto(Bytes))
```
Comment (benlangmuir) [inline clang/tools/driver/cc1_main.cpp]: Why are you getting the bytes here, they don't seem to be used.

File: llvm/cmake/modules/HandleLLVMOptions.cmake [hunk 4]
```diff
@@ -1342,6 +1342,12 @@ if (LLVM_CAS_ENABLE_CASID_OBJECT_OUTPUTS)
   append("-fcas builtin -fcas-builtin-path ${LLVM_CAS_BUILTIN_PATH_Default}" CMAKE_STATIC_LINKER_FLAGS)
 endif()
 
+set(LLVM_ENABLE_EXPERIMENTAL_CAS_BACKEND OFF CACHE BOOL "Use CASBackEnd")
```
Comment (benlangmuir) [inline llvm/cmake/modules/HandleLLVMOptions.cmake]: Nit: "Backend" is one word in this context, so shouldn't have uppercase E.

File: llvm/include/llvm/BinaryFormat/CAS.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  CAS_MEMPROT_READ = 1U << 0,
+  CAS_MEMPROT_WRITE = 1U << 1,
+  CAS_MEMPROT_EXEC = 1U << 2
+};
+
+enum class CASSymbolKind: uint8_t {
+  CAS_SYMBOL_STATIC = 0,
+  CAS_SYMBOL_UNDEF = 1,
+  CAS_SYMBOL_DYNAMIC = 2,
+  CAS_SYMBOL_FUTURE = 3
```
Comment (benlangmuir) [inline llvm/include/llvm/BinaryFormat/CAS.h]: Is "FUTURE" used anywhere?

**Review by benlangmuir — APPROVED** [review]
A few nitpicks, but nothing that would block merging if you prefer to merge first.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test
